### PR TITLE
Allow generated wrappers to subscribe/unsubscribe to events

### LINKED ISSUE #1869: Add support for generating subscribe/unsubscribe methods for contract events
Currently `abi-gen` does not provide any affordances for devs to subscribe/unsubscribe from contract events in the TS wrappers it generates. We want `abi-gen` to inspect the events emitted by a contract and provide `subscribe`/`unsubscribe` methods to listen to them.

### PR DESCRIPTION
**Pull request by xianny**
## Description

Provides a way to subscribe and unsubscribe to events in generated wrappers. This PR reproduces the functionality of the base ContractWrapper class in `@0x/contract-wrappers`.

Relevant code has been copied from `ContractWrapper` to a new class `SubscriptionManager` in `@0x/base-contract`, which is automatically instantiated in each generated wrapper. This allows us to:
* avoid breaking changes to constructors,
* preserve single responsibility of `BaseContract` class, and
* test `SubscriptionManager` as a separate component. 

`SubscriptionManager` is instantiated with type params which are generated from events defined in Solidity. 
Screen Shot 2019-07-18 at 9 32 23 AM

**I'm not sure about the name `SubscriptionManager`, open to suggestions.**

Resolves #1869

## Testing instructions

 I added a new test dummy contract in `@0x/abi-gen` that emits an event. 
* read generated code in `packages/abi-gen/test-cli/expected-output/typescript/event_dummy.ts`
* run unit tests

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/exchange/test/core.ts [hunk 1]
```diff
@@ -69,7 +70,7 @@ describe('Exchange core', () => {
     let exchange: ExchangeContract;
     let erc20Proxy: ERC20ProxyContract;
     let erc721Proxy: ERC721ProxyContract;
-    let erc1155Proxy: ERC721ProxyContract;
+    let erc1155Proxy: ERC1155ProxyContract;
```
Comment (feuGeneA) [inline contracts/exchange/test/core.ts]: An old oversight that was untested? (I saw there were multiple instances of this.)  Is it tested now?

File: packages/abi-gen-wrappers/src/generated-wrappers/dutch_auction.ts [hunk 2]
```diff
@@ -511,6 +511,7 @@ export class DutchAuctionContract extends BaseContract {
             return abiEncodedTransactionData;
         },
     };
+
```
Comment (feuGeneA) [inline packages/abi-gen-wrappers/src/generated-wrappers/dutch_auction.ts]: is this not a tslint violation? i know we've been saying "yes run prettier!" as part of abi-gen. but maybe we should also be running generated code through tslint too? :man_shrugging: maybe, maybe not... it seems like the right thing to do (but certainly not in this PR of course!) but it will likely create a not insignificant pile of work for us...

**Comment by feuGeneA** [comment]
In general LGTM, but I didn't "Approve" yet because I didn't yet review the new implementation that was copied in from `ContractWrappers`.

File: packages/abi-gen-wrappers/src/generated-wrappers/dutch_auction.ts (hunk 2 above)
Comment (xianny) [reply packages/abi-gen-wrappers/src/generated-wrappers/dutch_auction.ts]: This really bugged me! I ran it through `lint` and `prettier` but the whitespace line stayed. Our style rule enforces max of 1 line between method (I recall discussing this in an eng meeting a couple months ago, too!), but having **no** line between methods is **not** disallowed. Having no line between methods seems to be an `abi-gen-wrappers` thing, most other files have one line between methods. 

In any case, this whitespace is removed in the next commit. oops ^^

File: contracts/exchange/test/core.ts (hunk 1 above)
Comment (xianny) [reply contracts/exchange/test/core.ts]: yeah, these were weird. Maybe @hysz knows something about why they were written this way?

What's happening in this PR is that `*Contract` classes in `abi-gen-wrappers` are now typed (via the SubscriptionManager) to the type of events they have. Previously, the typescript compiler treated `ERC721ProxyContract` and `ERC1155ProxyContract` as interchangeable because they had identical interfaces. With `SubscriptionManager<ContractSpecificEventArgs, ContractSpecificEvent>`, they stop being interchangeable. Without this fix, the compiler complains about these lines later in the file:

```
erc1155ProxyWrapper = new ERC1155ProxyWrapper(provider, usedAddresses, owner);
erc1155Proxy = await erc1155ProxyWrapper.deployProxyAsync();
```

Comment (hysz) [reply contracts/exchange/test/core.ts]: Good find! Definitely a typo. I'm surprised this compiled/worked at all ??

This variable is later assigned as follows:
```
erc1155Proxy = await erc1155ProxyWrapper.deployProxyAsync();
```

Where, in `ERC1155ProxyWrapper`:
```
/**
     * @dev Deploys the ERC1155 proxy
     * @return Deployed ERC1155 proxy contract instance
     */
    public async deployProxyAsync(): Promise<ERC1155ProxyContract> {
        this._proxyContract = await ERC1155ProxyContract.deployFrom0xArtifactAsync(
            artifacts.ERC1155Proxy,
            this._provider,
            txDefaults,
        );
        this._proxyIdIfExists = await this._proxyContract.getProxyId.callAsync();
        return this._proxyContract;
    }
```

Comment (xianny) [reply contracts/exchange/test/core.ts]: Yeah, the compiler is happy if the variable instance has all the required fields of the type. I gather this is so that it can work with random js objects. Yikes.

File: packages/abi-gen-templates/contract.handlebars [hunk 3]
```diff
@@ -138,7 +144,88 @@ export class {{contractName}}Contract extends BaseContract {
...
+    public async subscribeAsync<ArgsType extends {{contractName}}EventArgs>(
+        eventName: {{contractName}}Events,
+        indexFilterValues: IndexedFilterValues,
+        callback: EventCallback<ArgsType>,
+        isVerbose: boolean = false,
+    ): Promise<string> {
+        assert.doesBelongToStringEnum('eventName', eventName, {{contractName}}Events);
+        assert.doesConformToSchema('indexFilterValues', indexFilterValues, schemas.indexFilterValuesSchema);
+        assert.isFunction('callback', callback);
+        const subscriptionManager = await this.getSubscriptionManagerAsync();
```
Comment (fabioberger) [inline packages/abi-gen-templates/contract.handlebars]: Let's make `getSubscriptionManagerAsync` into a private method.

File: packages/abi-gen/test-cli/test_typescript/test/event_test.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    describe('#subscribe', () => {
+        afterEach(async () => {
+            await eventDummy.unsubscribeAllAsync();
+            stubs.forEach(s => s.restore());
+            stubs = [];
+        });
+        it('Should allow unsubscribeAll to be called successfully after an error', (done: DoneCallback) => {
+            (async () => {
+                const callback = (err: Error | null, _logEvent?: DecodedLogEvent<EventDummyWithdrawalEventArgs>): any =>
+                    eventDummy.subscribeAsync(EventDummyEvents.Withdrawal, indexFilterValues, callback);
```
Comment (fabioberger) [inline packages/abi-gen/test-cli/test_typescript/test/event_test.ts]: What is going on here? What is after the `=>` in the line above?

File: packages/abi-gen/test-cli/test_typescript/test/event_test.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                stubs = [
```
Comment (fabioberger) [inline packages/abi-gen/test-cli/test_typescript/test/event_test.ts]: Where is `stubs` used?

File: packages/base-contract/src/index.ts [hunk 6]
```diff
@@ -146,6 +150,9 @@ export class BaseContract {
         }
         return rawEncoded;
     }
+    public getWeb3Wrapper(): Web3Wrapper {
```
Comment (fabioberger) [inline packages/base-contract/src/index.ts]: This exposes `web3-wrapper` on all contract public interfaces. The fact that we use `web3-wrapper` is an implementation detail and we don't want to surfaced to contract users. Why was this added?

File: packages/base-contract/src/types.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    InsufficientBalanceForTransfer = 'INSUFFICIENT_BALANCE_FOR_TRANSFER',
+    InsufficientEthBalanceForDeposit = 'INSUFFICIENT_ETH_BALANCE_FOR_DEPOSIT',
+    InsufficientWEthBalanceForWithdrawal = 'INSUFFICIENT_WETH_BALANCE_FOR_WITHDRAWAL',
+    InvalidJump = 'INVALID_JUMP',
+    OutOfGas = 'OUT_OF_GAS',
+    SubscriptionNotFound = 'SUBSCRIPTION_NOT_FOUND',
+    SubscriptionAlreadyPresent = 'SUBSCRIPTION_ALREADY_PRESENT',
+    ERC721OwnerNotFound = 'ERC_721_OWNER_NOT_FOUND',
+    ERC721NoApproval = 'ERC_721_NO_APPROVAL',
+    SignatureRequestDenied = 'SIGNATURE_REQUEST_DENIED',
```
Comment (fabioberger) [inline packages/base-contract/src/types.ts]: We only use `SubscriptionNotFound` and `SubscriptionAlreadyPresent` from there. Remove the rest.

File: packages/base-contract/src/utils/constants.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    OUT_OF_GAS_PATTERN: 'out of gas',
+    INVALID_TAKER_FORMAT: 'instance.taker is not of a type(s) string',
+    // tslint:disable-next-line:custom-no-magic-numbers
+    UNLIMITED_ALLOWANCE_IN_BASE_UNITS: new BigNumber(2).pow(256).minus(1),
+    DEFAULT_BLOCK_POLLING_INTERVAL: 1000,
+    ZERO_AMOUNT: new BigNumber(0),
+    ONE_AMOUNT: new BigNumber(1),
+    ETHER_TOKEN_DECIMALS: 18,
+    METAMASK_USER_DENIED_SIGNATURE_PATTERN: 'User denied transaction signature',
+    TRUST_WALLET_USER_DENIED_SIGNATURE_PATTERN: 'cancelled',
```
Comment (fabioberger) [inline packages/base-contract/src/utils/constants.ts]: Do we use all of these? Looks like a copy-pasta 🍝

File: packages/abi-gen-templates/contract.handlebars (hunk 3 above)
Comment (xianny) [reply packages/abi-gen-templates/contract.handlebars]: removing this one completely

File: packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts [hunk 9]
```diff
@@ -129,22 +131,8 @@ export class DutchAuctionWrapper extends ContractWrapper {
...
         // type assertions
         assert.doesConformToSchema('sellOrder', sellOrder, schemas.signedOrderSchema);
-        // get contract
-        const dutchAuctionInstance = await this._getDutchAuctionContractAsync();
         // call contract
-        const auctionDetails = await dutchAuctionInstance.getAuctionDetails.callAsync(sellOrder);
+        const auctionDetails = await this._dutchAuctionContract.getAuctionDetails.callAsync(sellOrder);
         return auctionDetails;
     }
-    private async _getDutchAuctionContractAsync(): Promise<DutchAuctionContract> {
```
Comment (xianny) [inline packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts]: I removed these `get__ContractAsync` methods everywhere for consistency. I think they are leftover from when instantiating `__Contract` involved something asynchronous. Now they can be instantiated directly in the constructor synchronously.

File: packages/base-contract/src/index.ts (hunk 6 above)
Comment (xianny) [reply packages/base-contract/src/index.ts]: This was added for unit tests. Simplifying the tests and getting rid of it.

Comment (fabioberger) [reply packages/base-contract/src/index.ts]: @xianny within tests, it's also kosher to do `(baseContract as any)._web3Wrapper`

File: packages/base-contract/src/subscription_manager.ts [hunk 10]
```diff
@@ -43,26 +35,24 @@ export abstract class ContractWrapper {
...
-        this._blockPollingIntervalMs =
-            blockPollingIntervalMs === undefined ? constants.DEFAULT_BLOCK_POLLING_INTERVAL : blockPollingIntervalMs;
         this._filters = {};
         this._filterCallbacks = {};
         this._blockAndLogStreamerIfExists = undefined;
         this._onLogAddedSubscriptionToken = undefined;
         this._onLogRemovedSubscriptionToken = undefined;
     }
-    protected _unsubscribeAll(): void {
+    public _unsubscribeAll(): void {
```
Comment (fabioberger) [inline packages/base-contract/src/subscription_manager.ts]: Now that `SubscriptionManager` is a internal class and not exposed to users of the wrappers, I'm fine with making these methods public on this class, but let's remove the underscore as that is meant to denote that the method is private.

File: packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts [hunk 11]
```diff
@@ -25,17 +25,17 @@ import {
...
  * the 0x V2 Coordinator extension contract.
  */
-export class CoordinatorWrapper extends ContractWrapper {
+export class CoordinatorWrapper {
     public abi: ContractAbi = Coordinator.compilerOutput.abi;
     public networkId: number;
     public address: string;
     public exchangeAddress: string;
     public registryAddress: string;
+    protected _web3Wrapper: Web3Wrapper;
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts]: Should be private.

File: packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts [hunk 12]
```diff
@@ -14,12 +14,11 @@ import { DutchAuctionWrapperError, OrderTransactionOpts } from '../types';
...
 import { _getDefaultContractAddresses } from '../utils/contract_addresses';
 
-import { ContractWrapper } from './contract_wrapper';
-
-export class DutchAuctionWrapper extends ContractWrapper {
+export class DutchAuctionWrapper {
     public abi: ContractAbi = DutchAuction.compilerOutput.abi;
     public address: string;
-    private _dutchAuctionContractIfExists?: DutchAuctionContract;
+    protected _web3Wrapper: Web3Wrapper;
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/dutch_auction_wrapper.ts]: Private. Protected means if a class inherits from `DutchAuctionWrapper`, it will be able to access `_webWrapper` which doesn't seem like what we want.

File: packages/contract-wrappers/src/contract_wrappers/erc20_proxy_wrapper.ts [hunk 13]
```diff
@@ -8,15 +8,14 @@ import * as _ from 'lodash';
...
-
 /**
  * This class includes the functionality related to interacting with the ERC20Proxy contract.
  */
-export class ERC20ProxyWrapper extends ContractWrapper {
+export class ERC20ProxyWrapper {
     public abi: ContractAbi = ERC20Proxy.compilerOutput.abi;
     public address: string;
-    private _erc20ProxyContractIfExists?: ERC20ProxyContract;
+    protected _web3Wrapper: Web3Wrapper;
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/erc20_proxy_wrapper.ts]: private.

File: packages/contract-wrappers/src/contract_wrappers/erc20_token_wrapper.ts [hunk 14]
```diff
@@ -20,35 +21,35 @@ import { assert } from '../utils/assert';
...
 /**
  * This class includes all the functionality related to interacting with ERC20 token contracts.
  * All ERC20 method calls are supported, along with some convenience methods for getting/setting allowances
  * to the 0x ERC20 Proxy smart contract.
  */
-export class ERC20TokenWrapper extends ContractWrapper {
+export class ERC20TokenWrapper {
     public abi: ContractAbi = ERC20Token.compilerOutput.abi;
     public UNLIMITED_ALLOWANCE_IN_BASE_UNITS = constants.UNLIMITED_ALLOWANCE_IN_BASE_UNITS;
+    protected _web3Wrapper: Web3Wrapper;
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/erc20_token_wrapper.ts]: private.

File: packages/contract-wrappers/src/contract_wrappers/erc721_token_wrapper.ts [hunk 15]
```diff
@@ -20,34 +21,34 @@ import { assert } from '../utils/assert';
...
 
 /**
  * This class includes all the functionality related to interacting with ERC721 token contracts.
  * All ERC721 method calls are supported, along with some convenience methods for getting/setting allowances
  * to the 0x ERC721 Proxy smart contract.
  */
-export class ERC721TokenWrapper extends ContractWrapper {
+export class ERC721TokenWrapper {
     public abi: ContractAbi = ERC721Token.compilerOutput.abi;
+    protected _web3Wrapper: Web3Wrapper;
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/erc721_token_wrapper.ts]: private. Please fix everywhere.
